### ISSUE TITLE
Alpine containers do not validate self-signed certs

### ISSUE THREAD

**Issue by enzian** [dotnet/corefx#41946]
When I build a containers that contains an AspNet Core app that accesses some other service via https, we get the error, that the cert cannot be validated:
```
System.Security.Authentication.AuthenticationException: The remote certificate is invalid according to the validation procedure.
...
```
Our Dockerfile is quite simple:
```
FROM mcr.microsoft.com/dotnet/core/aspnet:2.1-alpine AS runtime
COPY --from=build /app/identity/out /app
ENTRYPOINT ["dotnet", "/app/identity.dll"]
```
then we run the container using a compose file:
```
version: '3'
services:
  [...]
  identity:
    image: services.identity
    entrypoint: ""
    command: [ sh, -c, "update-ca-certificates && dotnet /app/identity.dll" ]
    networks:
      default:
        aliases: 
          - auth.hydrogen.com
    volumes:
      - ./configuration/ingress/cert.pem:/usr/local/share/cert.crt
    environment:
      ASPNETCORE_URLS: "https://*:3000"
    expose: 
      - "3000"
```
We access an upstream API that provides a self-signed cert, but this is the cert that is mounted to `/usr/local/share/cert.crt` which is appended to `/etc/ssl/certs/ca-certificates.crt` but it seems like in the current `aspnet:2.1-alpine` images, the certs in this file are not considered valid... Am I doing something wrong?

**Comment by davidsh** [comment]
>We access an upstream API that provides a self-signed cert, but this is the cert that is mounted to /usr/local/share/cert.crt which is appended to /etc/ssl/certs/ca-certificates.crt but it seems like in the current aspnet:2.1-alpine images, the certs in this file are not considered valid... Am I doing something wrong?

How did you generate the self-signed certificate? On Linux, there are additional requirements.  Unfortunately, the 'dotnet dev-certs' tool has a bug.  Please see this https://github.com/aspnet/AspNetCore/issues/7246

**Comment by enzian** [comment]
hey, that thread has gotten quite out of hand but here's my cert:
```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            7b:d9:....................................:e0:39
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN
        Validity
            Not Before: Oct 21 15:01:06 2019 GMT
            Not After : Oct 18 15:01:06 2029 GMT
        Subject: CN
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (2048 bit)
                Modulus:
                    00:b4:be:8a:23:82:3f:5e:44:f4:59:c6:51:13:b4:
                    0e:b5
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Basic Constraints:
                CA:TRUE
            X509v3 Key Usage:
                Digital Signature, Non Repudiation, Key Encipherment, Certificate Sign
            X509v3 Subject Alternative Name:
                DNS:*.sample.com, DNS:*.sample.web, DNS:sample.com, DNS:sample.web
    Signature Algorithm: sha256WithRSAEncryption
         75:7
         c:f3
-----BEGIN CERTIFICATE-----
MIIDrjCCA...npiEM8w==
-----END CERTIFICATE-----
```
considering that the `Certificate Sign` key usage is set. I'm not sure what I'm doing wrong here...

**Comment by davidsh** [comment]
>X509v3 Basic Constraints:
   CA:TRUE

That shouldn't be TRUE.  You want it to be FALSE in order for it to be used as a "leaf" node certificate. That might be why it doesn't work.

How exactly did you generate the self-signed certificate?  The 'dotnet dev-certs https' tool always generates a self-signed certificate with CA:FALSE. So, you must have used a different tool?

I would recommend you something like this:

<summary>
localhost.conf
</summary>
<details>
```
[req]
default_bits       = 2048
default_keyfile    = localhost.key
distinguished_name = req_distinguished_name
req_extensions     = req_ext
x509_extensions    = v3_ca

[req_distinguished_name]
countryName                 = Country Name (2 letter code)
countryName_default         = US
stateOrProvinceName         = State or Province Name (full name)
stateOrProvinceName_default = New York
localityName                = Locality Name (eg, city)
localityName_default        = Rochester
organizationName            = Organization Name (eg, company)
organizationName_default    = localhost
organizationalUnitName      = organizationalunit
organizationalUnitName_default = Development
commonName                  = Common Name (e.g. server FQDN or YOUR name)
commonName_default          = localhost
commonName_max              = 64

[req_ext]
subjectAltName = @alt_names

[v3_ca]
basicConstraints = CA:FALSE
subjectAltName = @alt_names
keyUsage       = keyCertSign, nonRepudiation, digitalSignature, keyEncipherment

[alt_names]
DNS.1   = localhost
DNS.2   = 127.0.0.1
```
</details>

And then use Openssl etc.:

```
sudo openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout localhost.key -out localhost.crt -config localhost.conf -passin pass:crypticpassword
```

**Comment by davidsh** [comment]
Also, I see that you are using a wildcard certificate

```
            X509v3 Subject Alternative Name:
                DNS:*.sample.com, DNS:*.sample.web, DNS:sample.com, DNS:sample.web
```

I know that there were problems with these kinds of certificates with .NET Core 2.1.

Have you tried .NET Core 3.0 to see those Alpine images work for you?

**Comment by enzian** [comment]
@davidsh thanks for the info, but so far I had no luck, here is what I do:

I have a compose file for testing that looks a little like this (truncated a few things):
```yaml

version: '3'

services:

  ingress:
    image: nginx:1.17-alpine
    volumes:
      - ./configuration/ingress/nginx.conf:/etc/nginx/nginx.conf:ro
      - ./configuration/ingress/ingress.key:/etc/nginx/tls.key:ro
      - ./configuration/ingress/ingress.crt:/etc/nginx/tls.crt:ro
    ports:
      - "443:443"
    expose:
      - "443"
    depends_on:
      - hydra
      - identity
    restart: unless-stopped

  hydra:
    image: oryd/hydra:v1
    command:
      serve all --config /etc/hydra.yaml
    volumes:
     - ./configuration/hydra/hydra.yaml:/etc/hydra.yaml:ro
     - ./configuration/hydra/server.key:/etc/tls/tls.key:ro
     - ./configuration/hydra/server.crt:/etc/tls/tls.crt:ro
    depends_on:
      - hydra-migrate
    restart: unless-stopped

  cockroachd:
    image: cockroachdb/cockroach:v2.1.6
    ports:
      - "26257:26257"
    command: start --insecure

  identity:
    build:
      context: ../H.Services.Identity
      dockerfile: Dockerfile
    image: h.services.identity
    entrypoint: ""
    command: [ sh, -c, "update-ca-certificates -v -f && dotnet /app/H.Services.Identity.dll" ]
    volumes:
      - ./configuration/identity/:/etc/identity/
      - ./configuration/ca/ca.crt:/usr/local/share/ca-certificates/ca.crt
    environment:
      ASPNETCORE_URLS: "https://*:3000"
      CONFIG_DIR: "/etc/identity/"
      ASPNETCORE_Kestrel__Certificates__Default__Path: /etc/identity/server.pfx
      ASPNETCORE_CONTENTROOT: /app/
    expose: 
      - "3000"
```
1. I generate the root ca certs and keys:
```ps
cd configuration/ca
openssl genrsa -out ca.key 4096
openssl req -x509 -new -nodes -key ca.key -sha512 -days 3650 -out ca.crt -extensions v3_ca -config ca.conf
```
the ca.conf file looks as follows:
```
[req]
default_bits        = 2048
distinguished_name  = req_distinguished_name
x509_extensions     = v3_ca
prompt = no

[req_distinguished_name]
C = ES
ST = MyState
L = MyCity
O = MyOrg
CN = sample.com

[ v3_ca ]
subjectKeyIdentifier = hash
authorityKeyIdentifier = keyid:always,issuer
basicConstraints = critical, CA:true
keyUsage = critical, digitalSignature, cRLSign, keyCertSign
```
2. I generate the TLS keys and certs for hydra and sign the csr with the root CAs key:
```
cd ../hydra
openssl genrsa -out server.key 4096
openssl req -new -key server.key -out server.csr -config csr.conf
openssl x509 -req -in server.csr -CA ../ca/ca.crt -CAkey ../ca/ca.key -CAcreateserial -out server.crt -days 10000 -extensions v3_ext -extfile csr.conf
```
3. I generate the keys and certs for the identity container:
```
cd ../identity
openssl genrsa -out server.key 4096
openssl req -new -key server.key -out server.csr -config csr.conf
openssl x509 -req -in server.csr -CA ../ca/ca.crt -CAkey ../ca/ca.key -CAcreateserial -out server.crt -days 10000 -extensions v3_ext -extfile csr.conf
cat server.crt | openssl x509 -text
openssl pkcs12 -export -out server.pfx -inkey server.key -in server.crt -certfile ../ca/ca.crt
```
like this, the issue persists using `mcr.microsoft.com/dotnet/core/aspnet:2.1-alpine`. I'll give `3.0-alpine` a go

**Comment by enzian** [comment]
It seems like when I use `mcr.microsoft.com/dotnet/core/aspnet:3.0-alpine` everything starts working when I follow the approach I described in [my previous comment](https://github.com/dotnet/corefx/issues/41946#issuecomment-544929504) which is awesome! But since `2.1` is LTS we should figure out why it fails there and not in 3.0

**Comment by davidsh** [comment]
>But since 2.1 is LTS we should figure out why it fails there and not in 3.0

I suspect it might be related to the use of wildcard certificates as I mentioned [here](https://github.com/dotnet/corefx/issues/41946#issuecomment-544614456). The fixes for that issue were not ported back to 2.1. Not all bug fixes are ported back to 2.1/2.2. There are always risks associated with back porting fixes. So, it depends on overall impact, number of customers affected, risk of fix, etc.

As a way of determining whether the use of wildcard certificates is impacting this, is there a way you test your scenario with using wildcard certificates?

**Comment by karelz** [comment]
Seems to be addressed in 3.0 and 3.1 (which is LTS - see [blog post](https://devblogs.microsoft.com/dotnet/announcing-net-core-3-1/)), let's close it unless there is strong signal that customers are running into this and cannot upgrade to 3.1.
